### PR TITLE
slot-based-collator: Implement dedicated block import

### PR DESCRIPTION
**Pull request by bkchr**
The `SlotBasedBlockImport` job is to collect the storage proofs of all blocks getting imported. These storage proofs alongside the block are being forwarded to the collation task. Right now they are just being thrown away. More logic will follow later. Basically this will be required to include multiple blocks into one `PoV` which will then be done by the collation task.

### CODE REVIEW

File: cumulus/polkadot-omni-node/lib/src/common/spec.rs [hunk 1]
```diff
@@ -92,6 +98,31 @@ fn warn_if_slow_hardware(hwbench: &sc_sysinfo::HwBench) {
...
+}
+
+pub(crate) struct ClientBlockImport;
+
+impl<Block: BlockT, RuntimeApi> InitBlockImport<Block, RuntimeApi> for ClientBlockImport
+where
+	RuntimeApi: Send + ConstructNodeRuntimeApi<Block, ParachainClient<Block, RuntimeApi>>,
+{
+	type BlockImport = Arc<ParachainClient<Block, RuntimeApi>>;
+	type ExtraReturnValue = ();
```
Comment (michalkucharczyk) [inline cumulus/polkadot-omni-node/lib/src/common/spec.rs:117]: nit: maybe `BlockImportAuxiliaryData` would be a better naming?
And also generic type in `ParachainService` could be `BIAuxiliaryData`?

File: cumulus/polkadot-omni-node/lib/src/common/spec.rs [hunk 2]
```diff
@@ -100,15 +131,28 @@ pub(crate) trait BaseNodeSpec {
...
 	>;
 
-	type BuildImportQueue: BuildImportQueue<Self::Block, Self::RuntimeApi>;
+	type BuildImportQueue: BuildImportQueue<
+		Self::Block,
+		Self::RuntimeApi,
+		<Self::WrapBlockImport as InitBlockImport<Self::Block, Self::RuntimeApi>>::BlockImport,
+	>;
+
+	type WrapBlockImport: InitBlockImport<Self::Block, Self::RuntimeApi>;
```
Comment (michalkucharczyk) [inline cumulus/polkadot-omni-node/lib/src/common/spec.rs:140]: nit: naming again: maybe `WrappedBlockImport`?

Comment (bkchr) [reply cumulus/polkadot-omni-node/lib/src/common/spec.rs:140]: It was first called `WrapBlockImport`, but then I decided for a different name. Apparently I failed to rename it everywhere :P Init in the context is more correct imo.

**Review by michalkucharczyk — APPROVED** [review]

File: cumulus/polkadot-omni-node/lib/src/common/spec.rs [hunk 3]
```diff
@@ -160,7 +204,10 @@ pub(crate) trait BaseNodeSpec {
 			.build(),
 		);
 
-		let block_import = ParachainBlockImport::new(client.clone(), backend.clone());
+		let (block_import, extra_return_value) =
```
Comment (michalkucharczyk) [inline cumulus/polkadot-omni-node/lib/src/common/spec.rs:207]:
```suggestion
		let (block_import, block_import_auxiliary_data) =
```

File: cumulus/polkadot-omni-node/lib/src/common/spec.rs [hunk 4]
```diff
@@ -178,7 +225,7 @@ pub(crate) trait BaseNodeSpec {
 			task_manager,
 			transaction_pool,
 			select_chain: (),
-			other: (block_import, telemetry, telemetry_worker_handle),
+			other: (block_import, telemetry, telemetry_worker_handle, extra_return_value),
```
Comment (michalkucharczyk) [inline cumulus/polkadot-omni-node/lib/src/common/spec.rs:228]:
```suggestion
			other: (block_import, telemetry, telemetry_worker_handle, block_import_auxiliary_data),
```

File: cumulus/polkadot-omni-node/lib/src/common/spec.rs [hunk 5]
```diff
@@ -213,7 +265,12 @@ pub(crate) trait NodeSpec: BaseNodeSpec {
 				let parachain_config = prepare_node_config(parachain_config);
 
 				let params = Self::new_partial(&parachain_config)?;
-				let (block_import, mut telemetry, telemetry_worker_handle) = params.other;
+				let (
+					block_import,
+					mut telemetry,
+					telemetry_worker_handle,
+					block_import_extra_return_value,
```
Comment (michalkucharczyk) [inline cumulus/polkadot-omni-node/lib/src/common/spec.rs:272]:
```suggestion
					block_import_auxiliary_data,
```

File: cumulus/polkadot-omni-node/lib/src/common/spec.rs [hunk 6]
```diff
@@ -343,6 +400,7 @@ pub(crate) trait NodeSpec: BaseNodeSpec {
 						announce_block,
 						backend.clone(),
 						node_extra_args,
+						block_import_extra_return_value,
```
Comment (michalkucharczyk) [inline cumulus/polkadot-omni-node/lib/src/common/spec.rs:403]:
```suggestion
						block_import_auxiliary_data,
```

File: cumulus/client/consensus/aura/src/collators/slot_based/block_import.rs [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+		{
+			let mut runtime_api = self.client.runtime_api();
+
+			runtime_api.set_call_context(CallContext::Onchain);
+
+			runtime_api.record_proof();
+			let recorder = runtime_api
+				.proof_recorder()
+				.expect("Proof recording is enabled in the line above; qed.");
+			runtime_api.register_extension(ProofSizeExt::new(recorder));
```
[the remark this answers is not in the document]
Comment (michalkucharczyk) [reply cumulus/client/consensus/aura/src/collators/slot_based/block_import.rs:111]: Yeah, maybe not. But after all it is a small code + expect string duplication. Up to you :)

**Review by lexnv — APPROVED** [review]
Not entirely familiar with the code, but looks good to me 🙏

**Pull request merged**